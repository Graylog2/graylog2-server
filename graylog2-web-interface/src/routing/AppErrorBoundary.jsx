import React from 'react';
import { withRouter } from 'react-router';
import PropTypes from 'prop-types';
import ErrorPage from '../pages/ErrorPage';

class AppErrorBoundary extends React.Component {
  static propTypes = {
    children: PropTypes.oneOfType([
      PropTypes.element,
      PropTypes.arrayOf(PropTypes.element),
    ]).isRequired,
    router: PropTypes.object.isRequired,
  };

  constructor(props) {
    super(props);
    this.state = {};
    this.unlisten = () => {};
  }

  componentDidMount() {
    const { router } = this.props;
    this.unlisten = router.listen(() => this.setState(() => ({ error: undefined, info: undefined })));
  }

  componentWillUnmount() {
    this.unlisten();
  }

  componentDidCatch(error, info) {
    this.setState({ error, info });
  }

  render() {
    const { error, info } = this.state;
    if (error) {
      return <ErrorPage error={error} info={info} />;
    }
    return this.props.children;
  }
}

export default withRouter(AppErrorBoundary);
