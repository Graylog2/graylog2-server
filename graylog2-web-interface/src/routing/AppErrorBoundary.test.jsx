import React from 'react';
import { mount } from 'enzyme';

import AppErrorBoundary from './AppErrorBoundary';

jest.mock('react-router', () => ({ withRouter: x => x }));
jest.mock('!style/useable!css!./NotFoundPage.css', () => ({ use: jest.fn(), unuse: jest.fn() }));

const suppressConsole = (fn) => {
  /* eslint-disable no-console */
  const originalConsoleError = console.error;
  console.error = () => {
  };

  fn();

  console.error = originalConsoleError;
  /* eslint-enable no-console */
};

const ErroneusComponent = () => {
  // eslint-disable-next-line no-throw-literal
  throw {
    message: 'Oh no, a banana peel fell on the party gorilla\'s head!',
    stack: 'This the stack trace.',
  };
};

const WorkingComponent = () => <div>Hello World!</div>;

describe('AppErrorBoundary', () => {
  it('registers to router upon mount', () => {
    const router = {
      listen: jest.fn(),
    };

    mount((
      <AppErrorBoundary router={router}>
        <WorkingComponent />
      </AppErrorBoundary>
    ));

    expect(router.listen).toHaveBeenCalled();
  });

  it('unregisters from router upon unmount', () => {
    const unlisten = jest.fn();
    const router = {
      listen: () => unlisten,
    };

    const wrapper = mount((
      <AppErrorBoundary router={router}>
        <WorkingComponent />
      </AppErrorBoundary>
    ));
    wrapper.unmount();

    expect(unlisten).toHaveBeenCalled();
  });

  it('displays child component if there is no error', () => {
    const router = {
      listen: jest.fn(),
    };
    const wrapper = mount((
      <AppErrorBoundary router={router}>
        <WorkingComponent />
      </AppErrorBoundary>
    ));

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(WorkingComponent)).toHaveLength(1);
  });

  it('displays error after catching', () => {
    const router = {
      listen: jest.fn(),
    };

    suppressConsole(() => {
      const wrapper = mount((
        <AppErrorBoundary router={router}>
          <ErroneusComponent />
        </AppErrorBoundary>
      ));

      const errorPage = wrapper.find('ErrorPage');
      expect(errorPage).toExist();
      expect(errorPage).toHaveProp('error', {
        message: 'Oh no, a banana peel fell on the party gorilla\'s head!',
        stack: 'This the stack trace.',
      });
      expect(errorPage).toHaveProp('info', { componentStack: '\n    in ErroneusComponent\n    in AppErrorBoundary (created by WrapperComponent)\n    in WrapperComponent' });
    });
  });

  it('resets error when navigation changes', () => {
    const router = {
      listen: jest.fn(),
    };

    suppressConsole(() => {
      const wrapper = mount((
        <AppErrorBoundary router={router}>
          <ErroneusComponent />
        </AppErrorBoundary>
      ));

      expect(wrapper).toIncludeText('banana peel');

      wrapper.setProps({ children: <WorkingComponent /> });
      const listenCallback = router.listen.mock.calls[0][0];
      listenCallback();
      // force update of component to reflect state changes
      wrapper.update();

      expect(wrapper.find(WorkingComponent)).toHaveLength(1);
      expect(wrapper).toIncludeText('Hello World');
    });
  });
});
