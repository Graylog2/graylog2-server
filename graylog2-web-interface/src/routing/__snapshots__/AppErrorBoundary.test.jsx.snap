// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppErrorBoundary displays child component if there is no error 1`] = `
<AppErrorBoundary
  router={
    Object {
      "listen": [MockFunction] {
        "calls": Array [
          Array [
            [Function],
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      },
    }
  }
>
  <WorkingComponent>
    <div>
      Hello World!
    </div>
  </WorkingComponent>
</AppErrorBoundary>
`;
